# <center> <b> <font color=Orange style="font-size:35px"> SuperStore US </font></b></center>

In [43]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

## Loading the Dataset

- The dataset is stored in an Excel file and consists of three sheets:Orders , Returns, Users


- These sheets are loaded into pandas DataFrames using the following code:

In [44]:
file_path = r'C:\Users\Alex\Desktop\SuperStoreUS_Assignment\SuperStoreUS.xlsx'

orders = pd.read_excel(file_path, sheet_name='Orders')
returns = pd.read_excel(file_path, sheet_name='Returns')
users = pd.read_excel(file_path, sheet_name='Users')

# 1.  `Sheet 'Orders'`

- To understand the structure of the Orders dataset, I inspect the first few rows:

In [45]:
print("Orders Data:")
orders.head()

Orders Data:


Row ID Order Priority  Discount  Unit Price  Shipping Cost  Customer ID  \
0   20847           High      0.01        2.84           0.93            3   
1   20228  Not Specified      0.02      500.98          26.00            5   
2   21776       Critical      0.06        9.48           7.29           11   
3   24844         Medium      0.09       78.69          19.99           14   
4   24846         Medium      0.08        3.28           2.31           14   

       Customer Name       Ship Mode Customer Segment Product Category  ...  \
0      Bonnie Potter     Express Air        Corporate  Office Supplies  ...   
1     Ronnie Proctor  Delivery Truck      Home Office        Furniture  ...   
2      Marcus Dunlap     Regular Air      Home Office        Furniture  ...   
3  Gwendolyn F Tyson     Regular Air   Small Business        Furniture  ...   
4  Gwendolyn F Tyson     Regular Air   Small Business  Office Supplies  ...   

    Region State or Province         City  Postal Code Order Date  Ship Date  \
0     West        Washington    Anacortes        98221 2015-01-07 2015-01-08   
1     West        California  San Gabriel        91776 2015-06-13 2015-06-15   
2     East        New Jersey      Roselle         7203 2015-02-15 2015-02-17   
3  Central         Minnesota   Prior Lake        55372 2015-05-12 2015-05-14   
4  Central         Minnesota   Prior Lake        55372 2015-05-12 2015-05-13   

      Profit Quantity ordered new    Sales Order ID  
0     4.5600                    4    13.01    88522  
1  4390.3665                   12  6362.85    90193  
2   -53.8096                   22   211.15    90192  
3   803.4705                   16  1164.45    86838  
4   -24.0300                    7    22.23    86838  

[5 rows x 25 columns]

- Additionally, I checked the dataset’s metadata, including column names, data types, and missing values:

In [46]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1952 entries, 0 to 1951
Data columns (total 25 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   Row ID                1952 non-null   int64         
 1   Order Priority        1952 non-null   object        
 2   Discount              1952 non-null   float64       
 3   Unit Price            1952 non-null   float64       
 4   Shipping Cost         1952 non-null   float64       
 5   Customer ID           1952 non-null   int64         
 6   Customer Name         1952 non-null   object        
 7   Ship Mode             1952 non-null   object        
 8   Customer Segment      1952 non-null   object        
 9   Product Category      1952 non-null   object        
 10  Product Sub-Category  1952 non-null   object        
 11  Product Container     1952 non-null   object        
 12  Product Name          1952 non-null   object        
 13  Product Base Margi

- The output shows that the dataset contains 1,952 rows and 25 columns with different data types

- The `Product Base Margin` column has missing values, which will be addressed in later steps

- Some numerical columns, such as `Customer ID`, `Order ID`, and `Postal Code`, are better stored as strings to avoid unintended mathematical operations. I converted them using:

In [47]:
orders['Customer ID'] = orders['Customer ID'].astype(str)
orders['Order ID'] = orders['Order ID'].astype(str)
orders['Postal Code'] = orders['Postal Code'].astype(str)

In [48]:
orders.columns

Index(['Row ID', 'Order Priority', 'Discount', 'Unit Price', 'Shipping Cost',
       'Customer ID', 'Customer Name', 'Ship Mode', 'Customer Segment',
       'Product Category', 'Product Sub-Category', 'Product Container',
       'Product Name', 'Product Base Margin', 'Country', 'Region',
       'State or Province', 'City', 'Postal Code', 'Order Date', 'Ship Date',
       'Profit', 'Quantity ordered new', 'Sales', 'Order ID'],
      dtype='object')

- To improve clarity, the column Quantity ordered new is renamed to Quantity ordered:

In [49]:
orders.rename(columns={'Quantity ordered new': 'Quantity ordered'}, inplace=True)

In [50]:
print("\nReturns Data:")
returns.head()


Returns Data:


Order ID    Status
0        65  Returned
1       612  Returned
2       614  Returned
3       678  Returned
4       710  Returned

- To gain further insights into the dataset's structure, I used:

In [51]:
returns.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1634 entries, 0 to 1633
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Order ID  1634 non-null   int64 
 1   Status    1634 non-null   object
dtypes: int64(1), object(1)
memory usage: 25.7+ KB


- The `Returns dataset` contains 1,634 rows and 2 columns:
- The `Order ID` column is currently stored as an integer (int64), but since it represents a categorical identifier rather than a numerical value, it should be converted to a string
- The `Status` column is already in the correct format (object)
- No missing values are present in this dataset

In [52]:
returns['Order ID']= returns['Order ID'].astype(str)

In [53]:
print("\nUsers Data:")
users.head()


Users Data:


Region  Manager
0  Central    Chris
1     East     Erin
2    South      Sam
3     West  William

In [54]:
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   Region   4 non-null      object
 1   Manager  4 non-null      object
dtypes: object(2)
memory usage: 196.0+ bytes


- During my earlier analysis of the `Orders dataset`, I noticed that the `Product Base Margin` column contains 16 missing values.

- I confirmed this by using:

In [55]:
orders.isna().sum()

Row ID                   0
Order Priority           0
Discount                 0
Unit Price               0
Shipping Cost            0
Customer ID              0
Customer Name            0
Ship Mode                0
Customer Segment         0
Product Category         0
Product Sub-Category     0
Product Container        0
Product Name             0
Product Base Margin     16
Country                  0
Region                   0
State or Province        0
City                     0
Postal Code              0
Order Date               0
Ship Date                0
Profit                   0
Quantity ordered         0
Sales                    0
Order ID                 0
dtype: int64

- To properly handle these missing values, we will need to decide on a strategy:

1. Remove rows with missing values (if they are insignificant to analysis).
2. Impute missing values using the mean, median, or, for example, replacing them with zeros

- After identifying 16 missing values in the `Product Base Margin` column, I chose the second approach to handle them: replacing missing values with the <strong>average margin for each Product Sub-Category</strong> rather than removing the rows

- To inspect the rows with missing values in `Product Base Margin`, I used:

In [56]:
# Rows with missing values in 'Product Base Margin'
missing_margin_rows = orders[orders['Product Base Margin'].isna()]
missing_margin_rows

Row ID Order Priority  Discount  Unit Price  Shipping Cost Customer ID  \
181    18261       Critical      0.06      276.20          24.49         335   
271    18305       Critical      0.01      128.24          12.65         508   
483    24764       Critical      0.09      349.45          60.00         868   
660    19185           High      0.09      349.45          60.00        1178   
700    20592         Medium      0.03      128.24          12.65        1237   
727    21848  Not Specified      0.08      128.24          12.65        1267   
734    22125            Low      0.10      238.40          24.49        1281   
736     4125            Low      0.10      238.40          24.49        1282   
977    22593           High      0.09      349.45          60.00        1739   
1228   19914  Not Specified      0.08       95.99          35.00        2211   
1507   19525       Critical      0.01      138.14          35.00        2660   
1729   23188           High      0.06      276.20          24.49        3048   
1774   20795       Critical      0.08      349.45          60.00        3119   
1860   18728  Not Specified      0.01      349.45          60.00        3257   
1935   23182       Critical      0.09      276.20          24.49        3380   
1943   19464  Not Specified      0.03       95.99          35.00        3388   

         Customer Name       Ship Mode Customer Segment Product Category  ...  \
181   Curtis O'Connell     Regular Air        Corporate        Furniture  ...   
271      Cameron Owens     Regular Air        Corporate        Furniture  ...   
483       Sharon Ellis  Delivery Truck        Corporate        Furniture  ...   
660         Sandy Hunt  Delivery Truck         Consumer        Furniture  ...   
700        Eva Simpson     Regular Air        Corporate        Furniture  ...   
727    Rosemary Branch     Regular Air        Corporate        Furniture  ...   
734     Pauline Denton     Regular Air   Small Business        Furniture  ...   
736        Dana Sharpe     Regular Air   Small Business        Furniture  ...   
977        Edna Pierce  Delivery Truck        Corporate        Furniture  ...   
1228        Anita Hahn     Express Air      Home Office  Office Supplies  ...   
1507      Jeffrey Page     Regular Air   Small Business  Office Supplies  ...   
1729     Tracy G Starr     Express Air        Corporate        Furniture  ...   
1774       Jay Hubbard  Delivery Truck        Corporate        Furniture  ...   
1860     Sharon Marcus  Delivery Truck         Consumer        Furniture  ...   
1935        Eva Decker     Regular Air      Home Office        Furniture  ...   
1943     Aaron Shaffer     Regular Air        Corporate  Office Supplies  ...   

       Region State or Province               City  Postal Code Order Date  \
181      West            Oregon            Medford        97504 2015-05-04   
271     South          Kentucky          Covington        41011 2015-04-18   
483   Central         Minnesota          Shoreview        55126 2015-03-06   
660     South           Florida  Altamonte Springs        32701 2015-04-09   
700   Central             Texas         Carrollton        75007 2015-01-31   
727     South           Florida         Boca Raton        33433 2015-05-12   
734   Central           Indiana          Vincennes        47591 2015-01-24   
736      East      Pennsylvania       Philadelphia        19134 2015-01-24   
977     South    North Carolina          Goldsboro        27534 2015-05-03   
1228     East          Maryland              Bowie        20715 2015-01-01   
1507     East             Maine             Gorham         4038 2015-02-24   
1729     West        California           Berkeley        94704 2015-03-05   
1774    South           Florida            Orlando        32839 2015-06-30   
1860     West        Washington           Longview        98632 2015-05-26   
1935    South           Georgia          La Grange        30240 2015-04-20   
1943     West      

- To analyze the missing data further, I extracted key columns:

In [57]:
missing_margin_rows[['Product Category', 'Product Sub-Category', 'Product Name', 'Sales', 'Profit']]

Product Category    Product Sub-Category  \
181         Furniture      Chairs & Chairmats   
271         Furniture      Chairs & Chairmats   
483         Furniture                  Tables   
660         Furniture                  Tables   
700         Furniture      Chairs & Chairmats   
727         Furniture      Chairs & Chairmats   
734         Furniture      Chairs & Chairmats   
736         Furniture      Chairs & Chairmats   
977         Furniture                  Tables   
1228  Office Supplies  Storage & Organization   
1507  Office Supplies  Storage & Organization   
1729        Furniture      Chairs & Chairmats   
1774        Furniture                  Tables   
1860        Furniture                  Tables   
1935        Furniture      Chairs & Chairmats   
1943  Office Supplies  Storage & Organization   

                                           Product Name    Sales     Profit  
181                            SAFCO Arco Folding Chair  3825.32  2639.4708  
271                         SAFCO Folding Chair Trolley   554.08   140.1354  
483   SAFCO PlanMaster Heigh-Adjustable Drafting Tab...  3918.98 -2946.0510  
660   SAFCO PlanMaster Heigh-Adjustable Drafting Tab...  2307.26  -369.1100  
700                         SAFCO Folding Chair Trolley  1145.60   790.4640  
727                         SAFCO Folding Chair Trolley   366.44  -379.3440  
734                     Safco Contoured Stacking Chairs  1774.50   875.2844  
736                     Safco Contoured Stacking Chairs  6654.39   460.6760  
977   SAFCO PlanMaster Heigh-Adjustable Drafting Tab...  5835.41   -90.7480  
1228                     Safco Industrial Wire Shelving   193.88  -425.2084  
1507              SAFCO Commercial Wire Shelving, Black   599.03  -321.5100  
1729                           SAFCO Arco Folding Chair  2610.56  1167.3800  
1774  SAFCO PlanMaster Heigh-Adjustable Drafting Tab...  3772.19   513.0840  
1860  SAFCO PlanMaster Heigh-Adjustable Drafting Tab...  5419.41  3739.3929  
1935                           SAFCO Arco Folding Chair  2941.42  3193.2840  
1943                     Safco Industrial Wire Shelving   894.64    67.0120

- Additionally, I checked which Product Sub-Categories contain missing values:

In [58]:
unique_values = orders[orders['Product Base Margin'].isna()]['Product Sub-Category'].unique()
unique_values

array(['Chairs & Chairmats', 'Tables', 'Storage & Organization'],
      dtype=object)

- All missing values in `Product Base Margin` belong to products from the `brand SAFCO`.
- To verify this, I filtered for rows where the product name contains `"SAFCO"`:

In [59]:
# Rows where the product is from SAFCO:
safco_rows = orders[orders['Product Name'].str.contains('SAFCO', case=False, na=False)]
len(safco_rows)


16

- The output confirmed that all SAFCO products lack Product Base Margin values, reinforcing the need to replace them with category-based averages rather than randomly chosen values

## <font color= 'Orange'>IMPUTING MISSING VALUES BASED ON PRODUCT SUB-CATEGORY AVERAGES</font>

- To determine appropriate replacement values, I computed the average `Product Base Margin` for each `Product Sub-Category`

In [60]:
# Average 'Product Base Margin' for each Product Sub-Category
subcategory_avg_margin = orders.groupby('Product Sub-Category')['Product Base Margin'].mean().round(2)

subcategory_avg_margin

Product Sub-Category
Appliances                        0.55
Binders and Binder Accessories    0.37
Bookcases                         0.66
Chairs & Chairmats                0.64
Computer Peripherals              0.60
Copiers and Fax                   0.43
Envelopes                         0.37
Labels                            0.38
Office Furnishings                0.52
Office Machines                   0.45
Paper                             0.38
Pens & Art Supplies               0.54
Rubber Bands                      0.52
Scissors, Rulers and Trimmers     0.64
Storage & Organization            0.67
Tables                            0.69
Telephones and Communication      0.59
Name: Product Base Margin, dtype: float64

- I will replace missing values in `Product Base Margin` column with averages for each Sub-Category:

- The following code was used to replace missing values in Product Base Margin:

In [61]:
# Imputing missing 'Product Base Margin' values based on the Product Sub-Category averages
orders.loc[orders['Product Base Margin'].isna(), 'Product Base Margin'] = orders['Product Sub-Category'].map(subcategory_avg_margin)

- To confirm that all missing values were successfully replaced, I ran the following check:

In [62]:
# Verifying the imputation
missing_after_imputation = orders['Product Base Margin'].isna().sum()
print(f"Number of missing values in 'Product Base Margin' after imputation: {missing_after_imputation}")

Number of missing values in 'Product Base Margin' after imputation: 0


- Now, we do not have missing values in `Product Base Margin` column

In [63]:
new_values = orders[orders['Product Name'].str.contains('SAFCO', case=False, na=False)]
new_values[['Product Sub-Category','Product Base Margin']]

Product Sub-Category  Product Base Margin
181       Chairs & Chairmats                 0.64
271       Chairs & Chairmats                 0.64
483                   Tables                 0.69
660                   Tables                 0.69
700       Chairs & Chairmats                 0.64
727       Chairs & Chairmats                 0.64
734       Chairs & Chairmats                 0.64
736       Chairs & Chairmats                 0.64
977                   Tables                 0.69
1228  Storage & Organization                 0.67
1507  Storage & Organization                 0.67
1729      Chairs & Chairmats                 0.64
1774                  Tables                 0.69
1860                  Tables                 0.69
1935      Chairs & Chairmats                 0.64
1943  Storage & Organization                 0.67

In [64]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1952 entries, 0 to 1951
Data columns (total 25 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   Row ID                1952 non-null   int64         
 1   Order Priority        1952 non-null   object        
 2   Discount              1952 non-null   float64       
 3   Unit Price            1952 non-null   float64       
 4   Shipping Cost         1952 non-null   float64       
 5   Customer ID           1952 non-null   object        
 6   Customer Name         1952 non-null   object        
 7   Ship Mode             1952 non-null   object        
 8   Customer Segment      1952 non-null   object        
 9   Product Category      1952 non-null   object        
 10  Product Sub-Category  1952 non-null   object        
 11  Product Container     1952 non-null   object        
 12  Product Name          1952 non-null   object        
 13  Product Base Margi

- With no missing values left, the dataset is now clean and ready for further transformations and analysis

### Checking if `Order Date` is greater than `Ship Date`

- To ensure consistency in shipping and ordering data, I checked whether any orders were placed after they were shipped:

In [65]:
order_after_ship = orders[orders['Order Date'] > orders['Ship Date']]
len(order_after_ship)

0

- Output confirms that all orders have valid dates, since `Order Date` always occurring before or on the same day as Ship Date.

- Then, I checked for cases where Profit exceeded Sales, which could indicate data inconsistencies:

In [66]:
profit_greater_than_sales = orders[orders['Profit'] > orders['Sales']]
len(profit_greater_than_sales)

70

- Since Profit should not exceed Sales, I filtered out these 70 inconsistent rows:

In [67]:
# Filter the dataset to keep only valid rows where Profit <= Sales
orders = orders[orders['Profit'] <= orders['Sales']]

### Checking for duplicates

- To ensure uniqueness in the dataset, I checked for duplicate rows:

In [68]:
#find the number of duplicates
orders.duplicated().sum()

0

- No duplicate entries were found

- Since the dataset includes a `Country` column, I verified if multiple countries were present:

In [69]:
orders['Country'].unique()

array(['United States'], dtype=object)

- Since all records belong to the United States, the `Country` column does not add any analytical value.
- I dropped this column, along with `Row ID`, which has no practical use:

In [70]:
orders.drop(['Row ID', 'Country'], axis=1, inplace=True)
orders.head()

Order Priority  Discount  Unit Price  Shipping Cost Customer ID  \
0           High      0.01        2.84           0.93           3   
1  Not Specified      0.02      500.98          26.00           5   
2       Critical      0.06        9.48           7.29          11   
3         Medium      0.09       78.69          19.99          14   
4         Medium      0.08        3.28           2.31          14   

       Customer Name       Ship Mode Customer Segment Product Category  \
0      Bonnie Potter     Express Air        Corporate  Office Supplies   
1     Ronnie Proctor  Delivery Truck      Home Office        Furniture   
2      Marcus Dunlap     Regular Air      Home Office        Furniture   
3  Gwendolyn F Tyson     Regular Air   Small Business        Furniture   
4  Gwendolyn F Tyson     Regular Air   Small Business  Office Supplies   

  Product Sub-Category  ...   Region State or Province         City  \
0  Pens & Art Supplies  ...     West        Washington    Anacortes   
1   Chairs & Chairmats  ...     West        California  San Gabriel   
2   Office Furnishings  ...     East        New Jersey      Roselle   
3   Office Furnishings  ...  Central         Minnesota   Prior Lake   
4  Pens & Art Supplies  ...  Central         Minnesota   Prior Lake   

  Postal Code Order Date  Ship Date     Profit Quantity ordered    Sales  \
0       98221 2015-01-07 2015-01-08     4.5600                4    13.01   
1       91776 2015-06-13 2015-06-15  4390.3665               12  6362.85   
2        7203 2015-02-15 2015-02-17   -53.8096               22   211.15   
3       55372 2015-05-12 2015-05-14   803.4705               16  1164.45   
4       55372 2015-05-12 2015-05-13   -24.0300                7    22.23   

   Order ID  
0     88522  
1     90193  
2     90192  
3     86838  
4     86838  

[5 rows x 23 columns]

- Finally, I rearranged the columns to improve data structure and analysis:

In [71]:
new_order = ['Order ID', 'Order Date', 'Ship Date','Ship Mode','Order Priority', 'Customer ID', 'Customer Name',  'Customer Segment',
             'Region', 'State or Province', 'City', 'Postal Code', 'Product Category', 'Product Sub-Category', 'Product Container',
            'Product Name', 'Product Base Margin', 'Sales', 'Quantity ordered', 'Discount', 'Unit Price', 'Shipping Cost','Profit']
orders=orders[new_order]
orders

Order ID Order Date  Ship Date       Ship Mode Order Priority  \
0       88522 2015-01-07 2015-01-08     Express Air           High   
1       90193 2015-06-13 2015-06-15  Delivery Truck  Not Specified   
2       90192 2015-02-15 2015-02-17     Regular Air       Critical   
3       86838 2015-05-12 2015-05-14     Regular Air         Medium   
4       86838 2015-05-12 2015-05-13     Regular Air         Medium   
...       ...        ...        ...             ...            ...   
1947    87536 2015-03-11 2015-03-12     Regular Air           High   
1948    87536 2015-03-11 2015-03-12     Regular Air           High   
1949    87534 2015-03-29 2015-03-31     Regular Air  Not Specified   
1950    87537 2015-04-04 2015-04-04     Express Air         Medium   
1951    87530 2015-02-08 2015-02-11     Express Air           High   

     Customer ID      Customer Name Customer Segment   Region  \
0              3      Bonnie Potter        Corporate     West   
1              5     Ronnie Proctor      Home Office     West   
2             11      Marcus Dunlap      Home Office     East   
3             14  Gwendolyn F Tyson   Small Business  Central   
4             14  Gwendolyn F Tyson   Small Business  Central   
...          ...                ...              ...      ...   
1947        3397        Andrea Shaw   Small Business  Central   
1948        3397        Andrea Shaw   Small Business  Central   
1949        3399        Marvin Reid   Small Business  Central   
1950        3400      Florence Gold   Small Business     East   
1951        3403      Tammy Buckley         Consumer     West   

     State or Province  ...          Product Sub-Category Product Container  \
0           Washington  ...           Pens & Art Supplies          Wrap Bag   
1           California  ...            Chairs & Chairmats        Jumbo Drum   
2           New Jersey  ...            Office Furnishings        Small Pack   
3            Minnesota  ...            Office Furnishings         Small Box   
4            Minnesota  ...           Pens & Art Supplies          Wrap Bag   
...                ...  ...                           ...               ...   
1947          Illinois  ...        Storage & Organization         Small Box   
1948          Illinois  ...  Telephones and Communication        Medium Box   
1949          Illinois  ...           Pens & Art Supplies        Small Pack   
1950     West Virginia  ...            Office Furnishings         Small Box   
1951           Wyoming  ...            Office Furnishings        Medium Box   

                                           Product Name Product Base Margin  \
0        SANFORD Liquid Accent™ Tank-Style Highlighters                0.54   
1        Global Troy™ Executive Leather Low-Back Tilter                0.60   
2     DAX Two-Tone Rosewood/Black Document Frame, De...                0.45   
3     Howard Miller 12-3/4 Diameter Accuwave DS ™ Wa...                0.43   
4                                            Newell 321                0.56   
...                                                 ...                 ...   
1947                                     Crate-A-Files™                0.59   
1948                       Bell Sonecor JB700 Caller ID                0.60   
1949                  Staples SlimLine Pencil Sharpener                0.60   
1950  Eldon Expressions Punched Metal & Wood Desk Ac...                0.57   
1951  Tenex 46" x 60" Computer Anti-Static Chairmat,...                0.65   

        Sales Quantity ordered  Discount  Unit Price  Shipping Cost     Profit  
0       13.01                4      0.01        2.84           0.93     4.5600  
1     6362.85               12      0.02      500.98          26.00  4390.3665  
2      211.15               22      0.06        9.48           7.29   -53.8096  
3     1164.45               16      0.09       78.69          19.99   803.4705  
4       22.23                7      0.08        3.28           2.31   -24.0300  

## Verifying Profit and Product Base Margin (PBM) Calculations

In [72]:
# Calculating new column Profit + Shipping'
orders['Profit + Shipping'] = orders['Profit'] + orders['Shipping Cost']

orders.describe()

Order Date                      Ship Date  \
count                           1882                           1882   
mean   2015-03-29 05:14:28.437831936  2015-03-31 04:17:51.200850176   
min              2015-01-01 00:00:00            2015-01-02 00:00:00   
25%              2015-02-11 00:00:00            2015-02-14 00:00:00   
50%              2015-03-28 00:00:00            2015-03-30 00:00:00   
75%              2015-05-15 00:00:00            2015-05-16 18:00:00   
max              2015-06-30 00:00:00            2015-07-08 00:00:00   
std                              NaN                            NaN   

       Product Base Margin         Sales  Quantity ordered     Discount  \
count          1882.000000   1882.000000       1882.000000  1882.000000   
mean              0.517837   1015.444957         13.065356     0.048842   
min               0.350000      2.250000          1.000000     0.000000   
25%               0.380000     61.505000          5.000000     0.020000   
50%               0.540000    212.020000         10.000000     0.050000   
75%               0.590000    833.802500         16.000000     0.070000   
max               0.850000  45737.330000        167.000000     0.210000   
std               0.137024   2601.309793         14.022367     0.031336   

        Unit Price  Shipping Cost        Profit  Profit + Shipping  
count  1882.000000    1882.000000   1882.000000        1882.000000  
mean    112.068571      13.213459     95.107270         108.320729  
min       1.140000       0.490000 -16476.838000      -16471.848000  
25%       6.480000       3.370000    -89.386500         -75.489500  
50%      22.550000       6.255000     -1.956850           2.774000  
75%     103.990000      14.520000     95.304600         106.549225  
max    6783.020000     164.730000   9228.225600        9248.215600  
std     400.278433      17.599782   1145.815369        1146.090571

- The `PBM` column contains <strong>only positive values</strong>, with a minimum of 0.35
- The `Profit + Shipping Cost` column, however, includes both positive and negative values, with a minimum of -16,471.85
- This suggests that the calculation method for PBM may differ from the expected formula

- To further investigate, I recalculated Total Costs and Profit using another approach, and then, I reviewed the key financial metrics:

In [73]:
orders['COGS'] = orders['Unit Price'] * orders['Quantity ordered']

orders['Total Costs'] = orders['Unit Price'] * orders['Quantity ordered'] + orders['Shipping Cost']

orders['Calculated Profit'] = orders['Sales'] - (orders['Unit Price'] * orders['Quantity ordered'] + orders['Shipping Cost'])

orders['Profit Difference'] = orders['Profit'] - orders['Calculated Profit']

orders['Calculated Product Base Margin'] = (orders['Sales'] - orders['Unit Price'] * orders['Quantity ordered']) / orders['Sales']

orders['PBM Difference'] = orders['Product Base Margin'] - orders['Calculated Product Base Margin']


orders[['Sales','Total Costs', 'COGS', 'Product Base Margin', 'Calculated Product Base Margin', 'PBM Difference','Profit', 'Calculated Profit', 'Profit Difference']].head()


Sales  Total Costs     COGS  Product Base Margin  \
0    13.01        12.29    11.36                 0.54   
1  6362.85      6037.76  6011.76                 0.60   
2   211.15       215.85   208.56                 0.45   
3  1164.45      1279.03  1259.04                 0.43   
4    22.23        25.27    22.96                 0.56   

   Calculated Product Base Margin  PBM Difference     Profit  \
0                        0.126826        0.413174     4.5600   
1                        0.055178        0.544822  4390.3665   
2                        0.012266        0.437734   -53.8096   
3                       -0.081231        0.511231   803.4705   
4                       -0.032839        0.592839   -24.0300   

   Calculated Profit  Profit Difference  
0               0.72             3.8400  
1             325.09          4065.2765  
2              -4.70           -49.1096  
3            -114.58           918.0505  
4              -3.04           -20.9900

- Significant differences exist between the original and calculated values of Profit and PBM.
- The recalculated Profit and PBM values are much lower than the originals.

- Since the recalculated PBM is lower than expected, it suggests that Unit Price may not represent the true cost price (COGS). 
- Therfore, I did one more check:

In [74]:
orders['Cost Price'] = (orders['Sales'] - orders['Sales'] * orders['Product Base Margin']) / orders['Quantity ordered']
orders[['Cost Price', 'Unit Price']]

Cost Price  Unit Price
0       1.496150        2.84
1     212.095000      500.98
2       5.278750        9.48
3      41.483531       78.69
4       1.397314        3.28
...          ...         ...
1947    4.722061       10.90
1948    2.602182        7.99
1949    4.798400       11.97
1950    3.892360        9.38
1951   35.455000      105.98

[1882 rows x 2 columns]

- From this, we can see that the product price we used to verify the PBM calculation (Unit Price) is higher than the actual price that was used (Cost Price).
- I assume that the `Unit Price` is not actually the base cost of the product (direct production costs) used in the PBM calculation, but rather includes some indirect costs, such as administrative costs, marketing and promotion expenses, etc.
- Accounting adjustments may also influence the reported Profit and PBM values

- Since we do not have detailed information on the exact cost structure or accounting adjustments used in the dataset, I will continue the analysis using the original values for Profit and PBM

- To clean up unnecessary columns, I dropped the temporary calculations:

In [75]:
orders.drop(['COGS','Total Costs','Calculated Product Base Margin', 'PBM Difference', 'Calculated Profit', 'Profit Difference', 'Profit + Shipping', 'Cost Price'], axis=1, inplace=True)
orders.head()

Order ID Order Date  Ship Date       Ship Mode Order Priority Customer ID  \
0    88522 2015-01-07 2015-01-08     Express Air           High           3   
1    90193 2015-06-13 2015-06-15  Delivery Truck  Not Specified           5   
2    90192 2015-02-15 2015-02-17     Regular Air       Critical          11   
3    86838 2015-05-12 2015-05-14     Regular Air         Medium          14   
4    86838 2015-05-12 2015-05-13     Regular Air         Medium          14   

       Customer Name Customer Segment   Region State or Province  ...  \
0      Bonnie Potter        Corporate     West        Washington  ...   
1     Ronnie Proctor      Home Office     West        California  ...   
2      Marcus Dunlap      Home Office     East        New Jersey  ...   
3  Gwendolyn F Tyson   Small Business  Central         Minnesota  ...   
4  Gwendolyn F Tyson   Small Business  Central         Minnesota  ...   

  Product Sub-Category Product Container  \
0  Pens & Art Supplies          Wrap Bag   
1   Chairs & Chairmats        Jumbo Drum   
2   Office Furnishings        Small Pack   
3   Office Furnishings         Small Box   
4  Pens & Art Supplies          Wrap Bag   

                                        Product Name Product Base Margin  \
0     SANFORD Liquid Accent™ Tank-Style Highlighters                0.54   
1     Global Troy™ Executive Leather Low-Back Tilter                0.60   
2  DAX Two-Tone Rosewood/Black Document Frame, De...                0.45   
3  Howard Miller 12-3/4 Diameter Accuwave DS ™ Wa...                0.43   
4                                         Newell 321                0.56   

     Sales Quantity ordered  Discount  Unit Price  Shipping Cost     Profit  
0    13.01                4      0.01        2.84           0.93     4.5600  
1  6362.85               12      0.02      500.98          26.00  4390.3665  
2   211.15               22      0.06        9.48           7.29   -53.8096  
3  1164.45               16      0.09       78.69          19.99   803.4705  
4    22.23                7      0.08        3.28           2.31   -24.0300  

[5 rows x 23 columns]

# 2. `Sheet 'Returns'`

- I  reviewed the dataset structure using:

In [76]:
returns.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1634 entries, 0 to 1633
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Order ID  1634 non-null   object
 1   Status    1634 non-null   object
dtypes: object(2)
memory usage: 25.7+ KB


- The dataset contains 1,634 rows, and no missing values were found.

- Another way to check if there are missing values:

In [77]:
returns.isna().sum()

Order ID    0
Status      0
dtype: int64

In [78]:
returns.duplicated().sum()

0

- No duplicate values were found.

- Checking unique values in the `Status` Column

In [79]:
returns['Status'].unique()

array(['Returned'], dtype=object)

- The `Status` column contains only one unique value: `"Returned"`.
- Since this column does not provide additional insight, it can be dropped.

In [80]:
returns.drop(['Status'], axis=1, inplace=True)
returns.head()

Order ID
0       65
1      612
2      614
3      678
4      710

# 3. `Sheet 'Users'`

In [81]:
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   Region   4 non-null      object
 1   Manager  4 non-null      object
dtypes: object(2)
memory usage: 196.0+ bytes


- The dataset contains only 4 rows with no missing values.

In [82]:
users.duplicated().sum()

0

- No duplicate values were found
- Since this dataset is very small and already clean, no additional preprocessing is needed

- Finally, I saved the cleaned datasets into new Excel file: `"Cleaned_SuperStoreUS.xlsx"`
- This file will be used for further analysis and creating interactive dashboard

In [83]:
output_file_xlsx = r'C:\Users\Alex\Desktop\SuperStoreUS_Assignment\Cleaned_SuperStoreUS.xlsx'

with pd.ExcelWriter(output_file_xlsx, engine='xlsxwriter') as writer:
    orders.to_excel(writer, sheet_name='Orders', index=False)
    returns.to_excel(writer, sheet_name='Returns', index=False)
    users.to_excel(writer, sheet_name='Users', index=False)


